presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -655,4 +661,88 @@ public String toString()
                     .toString();
         }
     }
+
+    /**
+     * This class can be used to make sure that only a single listener is added to the wrappedFuture.
+     * When the wrapped future is complete the listener is executed with the direct executor.
+     * OperatorContext synchronizes access to this class as this class is not thread safe and can be
+     * accessed by the memory revoking threads and the driver thread at the same time.
+     */
+    @NotThreadSafe
+    static class MemoryRevokingRequestedFuture<V>
+    {
+        private final AtomicReference<Runnable> listener = new AtomicReference<>();
+        private SettableFuture<V> wrappedFuture;
+
+        private MemoryRevokingRequestedFuture(SettableFuture<V> wrappedFuture)
+        {
+            this.wrappedFuture = requireNonNull(wrappedFuture, "wrappedFuture is null");
+        }
+
+        public static <V> MemoryRevokingRequestedFuture<V> create(SettableFuture<V> future)
+        {
+            MemoryRevokingRequestedFuture memoryRevokingRequestedFuture = new MemoryRevokingRequestedFuture<V>(future);
+            memoryRevokingRequestedFuture.setupListener();
+            return memoryRevokingRequestedFuture;
+        }
+
+        public boolean complete(@Nullable V value)
```

Comment:
As I discussed this code with @findepi we noticed here is race condition between `complete` and `setListener`. So you can miss actual unblocking on newly constructed composite future returned from `wakeUpOnRevokeRequest`.

Reply:
This particular problem can be (seemingly) fixed by re-consulting `MemoryRevokingRequestedFuture::isDone` _after_ listeners are registered (or reversing the order altogether within `wakeUpOnRevokeRequest`: first register new listeners and then check if any memory revoking futures is done).

(I am, however, not entirely convinced this as a whole is the cleanest way to solve the original problem... as commented earlier)